Kernel/VM/ScatterGatherList.h

```diff
@@ -7,29 +7,48 @@
 #pragma once
 
 #include <AK/Vector.h>
+#include <Kernel/Devices/BlockDevice.h>
 #include <Kernel/PhysicalAddress.h>
+#include <Kernel/VM/AnonymousVMObject.h>
 #include <Kernel/VM/MemoryManager.h>
 
 namespace Kernel {
 
-class ScatterGatherList {
-    struct ScatterGatherEntry {
+/// A Scatter-Gather List type that owns its buffers
+
+class ScatterGatherList : public RefCounted<ScatterGatherList> {
+public:
+    static NonnullRefPtr<ScatterGatherList> create(AsyncBlockDeviceRequest&, NonnullRefPtrVector<PhysicalPage> allocated_pages, size_t device_block_size);
+    const VMObject& vmobject() const { return m_vm_object; }
+    VirtualAddress dma_region() const { return m_dma_region->vaddr(); }
+    size_t scatters_count() const { return m_vm_object->physical_pages().size(); }
+
+private:
+    ScatterGatherList(AsyncBlockDeviceRequest&, NonnullRefPtrVector<PhysicalPage> allocated_pages, size_t device_block_size);
+    NonnullRefPtr<AnonymousVMObject> m_vm_object;
+    OwnPtr<Region> m_dma_region;
+};
+
+/// A Scatter-Gather List type that doesn't own its buffers
+
+class ScatterGatherRefList {
+    struct ScatterGatherRef {
         FlatPtr page_base;
```

Comment:
If we are using a buffer of someone else, we also own it. Why don't we use `ScatterGatherList` all the time?
In general, I'd prefer that we create a `ScatterGatherList` only when we need it like how it's done within the `AHCIPort`.
Is there a way to tell the VirtIO device to fetch data from a non-fixed place each time we need to get/send data from/to the device?